### PR TITLE
[Review only] noorders

### PR DESCRIPTION
**Pull request by proski**
This is not my code. I'm submitting this branch so that we have an option to examine the changes and comment on them now. It's a big change that requires careful review.

I want to make sure everybody is fine with the subpackage approach. It means that xrdp will be distributed with the libpainter code. There would be an option to distribute libpainter separately as well.

### CODE REVIEW

**Comment by proski** [comment]
Indeed, the change to `.travis.yml` was unneeded. I removed it.

File: libxrdp/libxrdp.c [hunk 1]
```diff
@@ -440,10 +461,26 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
-                                                     s, bpp,
-                                                     4096 - total_bufsize,
-                                                     i - 1, temp_s, e);
+
+                if (bpp > 24)
+                {
+                    LLOGLN(10, ("libxrdp_send_bitmap: 32 bpp"));
+                    lines_sending = xrdp_bitmap32_compress(data, width, height,
+                                                           s, 32,
+                                   (MAX_BITMAP_BUF_SIZE - 100) - total_bufsize,
```
Comment (proski) [inline libxrdp/libxrdp.c]: What is 100? Let's avoid magic constants in the new code.

File: libxrdp/libxrdp.c [hunk 2]
```diff
@@ -440,10 +461,26 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
 
                 p = s->p;
-                lines_sending = xrdp_bitmap_compress(data, width, height,
-                                                     s, bpp,
-                                                     4096 - total_bufsize,
-                                                     i - 1, temp_s, e);
+
+                if (bpp > 24)
+                {
+                    LLOGLN(10, ("libxrdp_send_bitmap: 32 bpp"));
```
Comment (proski) [inline libxrdp/libxrdp.c]: Never miss a change to print the actual value (bpp), not the assumed value (32).

File: libxrdp/libxrdp.c [hunk 3]
```diff
@@ -440,10 +461,26 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
+                    lines_sending = xrdp_bitmap32_compress(data, width, height,
+                                                           s, 32,
```
Comment (proski) [inline libxrdp/libxrdp.c]: 32 could be bpp

File: libxrdp/libxrdp.c [hunk 4]
```diff
@@ -481,54 +519,73 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
                     out_uint16_le(s, j); /* line size */
                     j = j * lines_sending;
                     out_uint16_le(s, j); /* final size */
+                    total_bufsize += 26;
```
Comment (proski) [inline libxrdp/libxrdp.c]: Magic 18, magic 26. Let's have names for those constants.

File: libxrdp/xrdp_bitmap_compress.c [hunk 5]
```diff
@@ -22,6 +22,8 @@
 
 #include "libxrdp.h"
 
+#define BC_MAX_BYTES (16 * 1024)
```
Comment (proski) [inline libxrdp/xrdp_bitmap_compress.c]: What is BC? Let's use COMPRESSED_MAX_BYTES

File: libxrdp/xrdp_bitmap_compress.c [hunk 6]
```diff
@@ -695,7 +697,7 @@ xrdp_bitmap_compress(char *in_data, int width, int height,
         out_count = end;
         line = in_data + width * start_line;
 
-        while (start_line >= 0 && out_count < 32768)
+        while (start_line >= 0 && out_count <= BC_MAX_BYTES)
```
Comment (proski) [inline libxrdp/xrdp_bitmap_compress.c]: Why did `<` become `<=` ? The code looks prone to buffer overflow.

File: libxrdp/xrdp_caps.c [hunk 7]
```diff
@@ -667,6 +677,17 @@ xrdp_caps_process_confirm_active(struct xrdp_rdp *self, struct stream *s)
         s->p = p + len + 4;
     }
 
+    if (self->client_info.no_orders_supported &&
+        (self->client_info.offscreen_support_level != 0))
+    {
+        g_writeln("xrdp_caps_process_confirm_active: not enough orders "
+                  "supported by client, client wants off screen bitmap but "
+                  "offscreen bitmaps disabled");
```
Comment (proski) [inline libxrdp/xrdp_caps.c]: Please write `offscreen` uniformly.

File: libxrdp/xrdp_caps.c [hunk 8]
```diff
@@ -872,9 +893,10 @@ xrdp_caps_send_demand_active(struct xrdp_rdp *self)
 
     /* INPUT_FLAG_SCANCODES 0x0001
        INPUT_FLAG_MOUSEX 0x0004
+       INPUT_FLAG_UNICODE 0x0010
        INPUT_FLAG_FASTPATH_INPUT 0x0008
        INPUT_FLAG_FASTPATH_INPUT2 0x0020 */
-    flags = 0x0001 | 0x0004;
+    flags = 0x0001 | 0x0004 | 0x0010;
```
Comment (proski) [inline libxrdp/xrdp_caps.c]: Why not just define those constants and use them in the expression?

File: libxrdp/xrdp_orders.c [hunk 9]
```diff
@@ -2235,6 +2235,11 @@ xrdp_orders_send_raw_bitmap(struct xrdp_orders *self,
 
     Bpp = (bpp + 7) / 8;
     bufsize = (width + e) * height * Bpp;
+    while (bufsize + 16 > MAX_ORDERS_SIZE)
```
Comment (proski) [inline libxrdp/xrdp_orders.c]: Magic number 16.

File: libxrdp/xrdp_orders.c [hunk 10]
```diff
@@ -2458,6 +2487,7 @@ xrdp_orders_send_raw_bitmap2(struct xrdp_orders *self,
     int pixel = 0;
     int e = 0;
 
+    g_writeln("xrdp_orders_send_raw_bitmap2:");
```
Comment (proski) [inline libxrdp/xrdp_orders.c]: Please use logging functions

File: libxrdp/xrdp_orders.c [hunk 11]
```diff
@@ -2479,6 +2509,11 @@ xrdp_orders_send_raw_bitmap2(struct xrdp_orders *self,
 
     Bpp = (bpp + 7) / 8;
     bufsize = (width + e) * height * Bpp;
+    while (bufsize + 14 > MAX_ORDERS_SIZE)
```
Comment (proski) [inline libxrdp/xrdp_orders.c]: Magic number 14

File: xrdp/xrdp_bitmap.c [hunk 12]
```diff
@@ -123,6 +123,12 @@ xrdp_bitmap_create(int width, int height, int bpp,
         self->data = (char *)g_malloc(width * height * Bpp, 0);
     }
 
+    if (self->type == WND_TYPE_SCREEN) /* noorders */
+    {
+        LLOGLN(0, ("xrdp_bitmap_create: noorders"));
+        self->data = (char *)g_malloc(width * height * Bpp, 0);
```
Comment (proski) [inline xrdp/xrdp_bitmap.c]: Please use `g_new` or at least add a space after the cast.

File: xrdp/xrdp_bitmap.c [hunk 13]
```diff
@@ -160,26 +166,27 @@ xrdp_bitmap_create_with_data(int width, int height,
...
+    {
+        case 8:
+            Bpp = 1;
+            break;
+        case 15:
+            Bpp = 2;
+            break;
+        case 16:
+            Bpp = 2;
+            break;
```
Comment (proski) [inline xrdp/xrdp_bitmap.c]: We need two inline functions - sparse_Bpp (returning 1, 2, 3, 4) and aligned_Bpp (returning 1, 2, 4). Repeating the same code is not useful.

File: libxrdp/xrdp_caps.c [hunk 14]
```diff
@@ -159,6 +160,15 @@ xrdp_caps_process_order(struct xrdp_rdp *self, struct stream *s,
     DEBUG(("desktop cache size %d", i));
     in_uint8s(s, 4); /* Unknown */
     in_uint8s(s, 4); /* Unknown */
+
+    /* check if libpainter should be used for drawing, instead of orders */
+    if (!(order_caps[TS_NEG_DSTBLT_INDEX] && order_caps[TS_NEG_PATBLT_INDEX] &&
+          order_caps[TS_NEG_SCRBLT_INDEX] && order_caps[TS_NEG_MEMBLT_INDEX]))
+    {
+        g_writeln("xrdp_caps_process_order: not enough orders supported by client, using painter.");
```
Comment (proski) [inline libxrdp/xrdp_caps.c]: Let's use uniform logging,

File: xrdp/xrdp_wm.c [hunk 15]
```diff
@@ -1490,6 +1457,80 @@ xrdp_wm_key(struct xrdp_wm *self, int device_flags, int scan_code)
...
+    }
+
+    for (index = 8; index < 256; index++)
+    {
+        if (unicode == self->keymap.keys_shift[index].chr)
+        {
+            if (device_flags & KBD_FLAG_UP)
+            {
+                xrdp_wm_key(self, device_flags, index - 8);
+                xrdp_wm_key(self, KBD_FLAG_UP, 42);
```
Comment (proski) [inline xrdp/xrdp_wm.c]: Magic number 42 ;)

File: xrdp/xrdp_wm.c [hunk 16]
```diff
@@ -1721,6 +1762,9 @@ callback(long id, int msg, long param1, long param2, long param3, long param4)
         case 4: /* RDP_INPUT_SCANCODE */
             rv = xrdp_wm_key(wm, param3, param1);
             break;
+        case 5: /* RDP_INPUT_UNICODE */
+            rv = xrdp_wm_key_unicode(wm, param3, param1);
+            break;
```
Comment (proski) [inline xrdp/xrdp_wm.c]: It looks like the keyboard changes are independent and can be submitted separately.

**Comment by jsorg71** [comment]
I'm planning to break this up and make new PRs.

**Comment by proski** [comment]
Rebased on top of the current devel branch, Unicode input has been merged already.

File: libxrdp/xrdp_caps.c [hunk 17]
```diff
@@ -85,6 +85,7 @@ xrdp_caps_process_general(struct xrdp_rdp *self, struct stream *s,
     /* op2 is a boolean to use compact bitmap headers in bitmap cache */
     /* set it to same as 'use rdp5' boolean */
     self->client_info.op2 = self->client_info.use_compact_packets;
+    self->client_info.op1 = extraFlags & NO_BITMAP_COMPRESSION_HDR; /* NO_BITMAP_COMPRESSION_HDR */
     /* FASTPATH_OUTPUT_SUPPORTED 0x0001 */
```
Comment (proski) [inline libxrdp/xrdp_caps.c]: Please remove useless comments

**Comment by proski** [comment]
Closing, it's out of date

**Pull request closed without merging**